README.md
# Project Cynical
A Card Game bot.

It was made after a few members of the Creosphere Discord server decided to play a bit of some sort of a twist of Pokemon and Yu-Gi-Oh. I decided to code it into a bot because why not :le:


Note: You are not permitted to self-host. This was never intended to be a public bot.


changelog
***Pre-Release 0 - The Creation***
-> Implemented the first foundation of Card Battle
-> Implemented 6 first Universal Cards and 2 first Special Cards
-> Implemented all/most needed commands
-> Implemented emote response

**Pre-Release 1**
-> Added alias for `c.shutdown`
-> Implemented changelog (I need to stop implementing changelogs after release)

***Pre-Release 2***
-> Implemented and modified 2 new Cards: **Poison Target** and **Healing Potion**
-> Implemented additional effect of Protection-inducing Cards' where successive use halves the Card's base accuracy (this was somehow forgotten, whoops)
-> Implemented additional effect of Card "Jumpscare" where it fails if attempted to be used successively (this was also forgotten, oopsies)
-> Modified foundation of Card Battle
-> Modified changelog
-> Reset highscores.list to default values
-> Fixed a bug that `playerFlinched` variable is not reset to default values when resetting variables after a battle

**Pre-Release 3**
-> Fixed a nasty bug which user's HP drops to exactly 7 HP after poison has been cured after a number of turns
-> Made many changes to the outcome messages and some texts to prevent confusion with same displayed names but different usernames/discriminators

**Pre-Release 4**
-> Fixed an exploitable bug which Healing Potion can theoretically heal the user beyond the maximum HP
-> Made yet many more changes to the outcome messages

***Alpha 1.0 - QoL Update 1***
[Suggested by SanskariHydra]
-> Added Cards' respective energy costs besides the Cards' names when using `c.list`
-> Implemented `c.resign` which ends the battle with an opposition victory.
[Self-Implemented]
-> Implemented and modified a new Universal Card: **Warn**
-> Introduced and implemented a new Special Card: **The Eclipse** (ID: 1003)
-> Implemented the ability to select a Card by the position of said Card in the user's deck of Cards
-> Implemented `c.changelog` (yet another forgotten feature)
-> Buffed the accuracy stat change of Card "Potion of Invisibility" (+10% instead of +5%)
-> Increased chance of getting a Special Card if the player does not have a Special Card specific to them (20% instead of 10%)
-> The defending player now has 20% chance of regaining 2 Energy instead of just 1 Energy
-> The battle summary screen now shows the Cards' positions in each player's decks
-> Updated `c.use` and the message that gets sent at the start of any player's turn
-> Fixed a bug (accidentally executed by SanskariHydra#7785) which the bot suddenly displays a message while it is preparing to shutdown
-> Fixed an unknown potential bug

***Alpha 1.1 - QoL Update 2***
-> Implemented the ability for the user to see their own scores in `c.leaderboards` if their total scores are not in the Top 10
-> Added `c.changelog` in the help embed
-> Added optional argument for `c.shutdown`
-> Split the changelog for Alpha 1.2 because Discord complained about it being too long to fit in
-> Fixed a series of incredibly notorious bugs involving the ability to select a Card by its position in the player's deck
-> Fixed yet another series of bugs regarding the leaderboards embed showing the high scores incorrectly and other bugs
-> Fixed a bug that `c.resign` didn't work
-> Fixed a bug that `c.leaderboards` would either not display the high scores embed or just display some random string characters

***Beta 1.0 - Rewritten Update 1***
-> Rewritten the code, not necessarily from the ground up though (lol)
-> Introduced and implemented two new Special Cards: **Guard Break** and **Limit Break**
-> Implemented and modified three new Universal Cards: **Mute**, **Ban**, and **Push**
-> Implemented Card Classifications (Offensive, Defensive, Miscellaneous) so you can be sure you always have at least one of each
-> Implemented support for multiple battles -- but not in the same channels (Inspired by the late Pokecord)
-> Implemented an automatic battle stopper when both parties skipped 2 turns each (a total of 4 turns) in a row (Inspired by Poketwo)
-> Increased the chance of getting a Special Card (40% instead of 20%)
-> Increased the number of Cards a player's Deck can have (8 instead of 6)
-> The defending player now has 10% chance of regaining 3 Energy instead of 1 or 2 Energy
-> Merged the changelog for Alpha 1.2
-> Greatly enhanced the `c.help` command
=-> Grouped together the commands in appropriate categories
=-> Implemented support for more information about a specific command
-> Redesigned the `c.moveinfo` and the battle summary embeds
-> Added aliases for most commands
-> Greatly modified the changelog
-> Removed the bot's emote response (wasn't really necessary)
-> Fixed the energy costs with Cards **Psychic Protect** and **The Ban Hammer**
-> Fixed the chances that the second player would get a Special Card (it was 10% instead of 20%/40%, sorry)
-> Fixed many bugs that have been encountered during the program rewrite

**Beta 1.1 - Small Update 1**
-> Introduced and implemented a new Universal Card: **Early Assault**
-> Made the battle summary screen to show the current turn of the battle
-> Changed `c.skip` to `c.restore`, with an implemented energy restoration effect for the player [Suggested by MarwanPGX]
=-> There is a chance that the player can regenerate 4 to 5 Energy as opposed to a maximum of 3 Energy normally
-> Made the bot's command prefix case insensitive [Thanks CriticalElement :)]
-> Slightly modified the changelog
-> Fixed a glitch which the chances of a Special Card added on one's deck is much lower than what was intended
=-> Because of this fix, the chances of a Special Card are now reduced (25% instead of 40%)
-> Fixed a bug which using the **Healing Potion** Card while not being poisoned does not show the difference
-> Fixed a bug that the footer displays the wrong avatar and username after finishing a turn

**Beta 1.11 - Quick Update 1**
-> Increased maximum HP to 200
-> Fixed **Early Assault** displaying the wrong Card Power in `c.cardinfo`


cardDefines.py
from cards import Card, SpecialCard
from constants import UserIDs

offensiveCards = {
    "punch": Card(0, "Punch", 15, 95, 1, "The basic attack.", "No additional effects.", "offensive"),
    "super punch": Card(1, "Super Punch", 30, 75, 3, "Deals more damage than a regular punch.", "Increased critical hit chance.", "offensive"),
    "double slap": Card(3, "Double Slap", 15, 75, 3, "\"Slap like now and subscribe!\" - a certain Italian bassist", "Hits twice. Each slap has a chance of flinching the target.", "offensive"),
    "push": Card(11, "Push", 0, 80, 4, "\"Do you believe in gravity?\" - A JoJo reference", "Has a chance of damaging the target, if the Card's accuracy check passed. Unaffected by Protection. Damage cannot be increased by critical hit.", "offensive,miscellaneous"),
    "early assault": Card(12, "Early Assault", 30, 65, 5, "(Placeholder text here.)", "The Card's base Power doubles in the first 4 total turns of the battle, and halves by Turn 8 (total). Bound to fail by Turn 12 (total) onwards.", "offensive"),
    "snipe shot": Card(13, "Snipe Shot", 35, 60, 7, "Gottem! Wait... oh.", "Increased critical hit damage.", "offensive"),
    "chain breaker": Card(15, "Chain Breaker", 15, 100, 4, "```\nUser 3: lol\nUser 4: lol\nUser 69: :le:```", "Breaks the target's chain after the move, decreasing Defense equal to target's chain level. Power is multiplied by 75% for each chain level. Fails if the target does not have a chain.", "offensive"),
    "fissure": Card(17, "Fissure", 999, 30, 8, "Cracks open the ground below the target, potentially killing them.", "No additional effects.", "offensive")
}

defensiveCards = {
    "potion of invisibility": Card(5, "Potion of Invisibility", -1, 100, 3, "Makes you invisible.", "Increases evasion by 5%.", "defensive", True),
    "chain": Card(14, "Chain", -1, 80, 3, "```\nUser 1: lol\nUser 2: lol\nUser 3: lol```", "Increases user's Defense by 10%. For each successive use, elevates chain level by 1.", "defensive", True),
    "protect": Card(18, "Protect", -1, 85, 5, "\"I protect the hooman\" - cat", "Greatly reduces an opponent's attack damage by 85%. If it's an OHKO attack, it only reduces the user's HP to 1. If it's a status-applying Card, it cancels it. Lasts one turn. Successive use halves the Card's accuracy. Fails if the user has any status effect.", "defensive", True),
    "banishing shield": Card(19, "Banishing Shield", -1, 40, 7, "!ban @target, but for self-defense", "Any Card used on the user will automatically fail and be Disabled for use for the rest of the battle. This Card can only be used once, and lasts 2 turns only. Fails if both the user and the target have not used any prior Card or when user has any status effect.", "defense,miscellaneous", True)
}

SupportCards = {
    "poison target": Card(6, "Poison Target", -1, 70, 4, "Give them poisoned food or drink, see what happens.", "Poisons the target for 2 - 5 turns. (Target loses HP by 5% at the end of a turn.) Fails if the target has any active status effect.", "support,offensive"),
    "healing potion": Card(7, "Healing Potion", -1, 100, 2, "Buy the best, all-cure Healing Potion™ for only $69", "Heals the user by 45%. If the user is poisoned, has a chance to cure the user of poisoning. Successive use reduces the Card's accuracy by 25%.", "support", True),
    "revival": Card(16, "Revival", -1, 60, 7, "lol let's make chat alive again.", "Allows the user to revive and be healed by 50% HP should the user get KO'ed. Fails while user's **Revivable** status is active.", "support,defensive", True),
    "restore": Card(1000, "Restore", -1, 100, 0, "Wait, that's illegal", "Restores the User's energy. Alias: `c.restore`", "miscellaneous,support", True)
}

MiscCards = {
    "splash": Card(2, "Splash", -1, 80, 2, "Nothing happens... or was there?", "If the Card hits, has a chance of either flinching the target or dealing 1 HP damage to the target.", "miscellaneous"),
    "jumpscare": Card(4, "Jumpscare", -1, 80, 3, ":)", "Flinches the target. Also decreases the target's accuracy by 5%, but increases target's evasion by 10%. Fails when used in succession, when either of the target's Accuracy or Evasion can no longer be changed, or when user is currently Blocked.", "miscellaneous"),
    "warn": Card(8, "Warn", -1, 60, 5, "!warn @target", "Prevents the target from using a Card for 2 turns. (Inflicts the target with the \"Blocked\" status effect.)", "miscellaneous"),
    "mute": Card(9, "Mute", -1, 50, 6, "!tempmute @target 4d", "Forbids the target from using the previous Card again for 2 - 5 turns. Bypasses all forms of protection. Fails if the target hasn't used a Card at least once. Successive use reduces the Card's accuracy by 75%", "miscellaneous"),
    "banishing slam": Card(10, "Banishing Slam", -1, 40, 7, "!ban @target", "Disables the last Card used by the target for the duration of the whole battle, and disables this Card from being re-used by user again for the same duration. Fails if target hasn't used a Card at least once. **Does not actually deal damage.**", "miscellaneous")
}

SpecialCards = {
    "psychic protect": SpecialCard(1001, "Psychic Protect", -1, 60, 6, "lmao ok.", "Fully protects the user from almost all attacks for one move. Successive use halves the Card's accuracy. Fails when the user is currently poisoned.", UserIDs["navyblue"], "Accuracy is boosted to 70% and cures poisoning. If the user is not poisoned, adds an extra turn of protection instead.", "defensive,support", True),
    "the ban hammer": SpecialCard(1002, "The Ban Hammer", 999, 30, 9, "begone.", "Guaranteed to bypass **Revival**, **Protection**, and **Disabling Aura**, and has a chance to bypass **Psyched Protection**. For each miss or fail, increase user's accuracy by 5%.", UserIDs["theAstra"], "Accuracy is boosted to 55% and Energy cost is reduced to 7 Energy. Increases user's accuracy by 10% instead if the move fails or misses.", "offensive"),
    "the eclipse": SpecialCard(1003, "The Eclipse", -1, 70, 5, "The worlds align and the eclipse falls, spreading darkness around the world.", "Reduces the accuracy of all players in battle (opponent's: 25 / user's: 15) and the user is healed by 40% the maximum HP. Successive use reduces the Card's accuracy by 25%. Fails if either player's accuracy can no longer be deducted.", UserIDs["SanskariHydra"], "Opponent's accuracy is reduced by 30 and user's accuracy is reduced by 10 instead. User's heals by double the percent any other user heals (70% instead of 35%).", "defensive,support"),
    "guard break": SpecialCard(1004, "Guard Break", -1, 100, 7, "I'm weak.", "Maximizes user's Evasion but lowers Defense by 70. Fails if user's stats cannot be lowered or raised further. Once used, this Card can no longer be used for the rest of the battle (essentially Disabling itself).", 0, "No additional effects.", "defensive,miscellaneous", True),
    "limit break": SpecialCard(1005, "Limit Break", -1, 100, 10, "UNLIMITED POWER!!!", "Breaks the maximum limit of the user's stats, allowing for abnormally high stats. Once used, this Card can no longer be used for the rest of the battle (essentially Disabling itself). Fails if the User hasn't used a prior Card.", 0, "No additional effects.", "miscellaneous", True)
}

UniversalCards = []
SpecialCards_ = []
AllCards = dict()
NormalCards = dict()

def initUniversalCards():
    global UniversalCards
    
    # 1 - 10
    UniversalCards.append("punch")
    UniversalCards.append("super punch")
    UniversalCards.append("splash")
    UniversalCards.append("double slap")
    UniversalCards.append("jumpscare")
    UniversalCards.append("potion of invisibility")
    UniversalCards.append("poison target")
    UniversalCards.append("healing potion")
    UniversalCards.append("warn")
    UniversalCards.append("mute")
    
    # 11 - 20
    UniversalCards.append("banishing slam")
    UniversalCards.append("push")
    UniversalCards.append("early assault")
    UniversalCards.append("snipe shot")
    UniversalCards.append("chain")
    UniversalCards.append("chain breaker")
    UniversalCards.append("revival")
    UniversalCards.append("fissure")
    UniversalCards.append("protect")
    UniversalCards.append("banishing shield")
    
    # 1000 should not be included. 

def initSpecialCards():
    global SpecialCards_
    x = 0
    while x < 1001:
        SpecialCards_.append(None)
        x += 1
        
    SpecialCards_.append("psychic protect")
    SpecialCards_.append("the ban hammer")
    SpecialCards_.append("the eclipse")
    SpecialCards_.append("guard break")
    SpecialCards_.append("limit break")

def initAllCards():
    global offensiveCards, defensiveCards, MiscCards, SpecialCards, AllCards, NormalCards
    
    NormalCards = {**offensiveCards, **defensiveCards, **SupportCards, **MiscCards}
    AllCards = {**NormalCards, **SpecialCards}
    